base/sort.jl

```diff
@@ -992,7 +994,13 @@ julia> v
  2
 ```
 """
-sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...)
+function sort(v; kws...)
+    IteratorSize(v) == HasShape{0}() && throw(ArgumentError("$v cannot be sorted"))
```

Comment:
Shouldn't infinite (and maybe unknown) also be errors? I see that people probably want an error for `sort(1)`, so ok, but kind of strange that that of all things would be disallowed.

Reply:
Unknown sizes should be supported
```julia
sort(w for w in ["karma", "infracostalis", "postencephalon", "Elia"] if all(islowercase, w))
```

For almost all infinite iterators, we already throw inside `copymutable`, but perhaps someone could define an infinite iterator that can be `copymutable`ed, returning a sparse vector of infinite length. If they also define a sorting method for that sparse representation, then it would be a mistake to throw on infinite iterators.

Reply:
But `sort` is an eager operation - surely if someone wants to sort their infinite iterator lazily, they have to implement it either way and not just rely on the generic fallback that's supposed to collect all elements of the iterator?

Reply:
For generic iterables, `copymutable` `collect`s, but it only needs to make a mutable copy, not necessarily instantiate every element. For example, I believe this is allowed:
```julia
struct WithTrailingZerosIterable
    head::Int
    tail::Union{Nothing, WithTrailingZerosIterable}
end

Base.iterate(x::WithTrailingZerosIterable) = iterate(x, x)
Base.iterate(::WithTrailingZerosIterable, x::WithTrailingZerosIterable) = x.head, x.tail
Base.iterate(::WithTrailingZerosIterable, ::Nothing) = 0, Nothing

Base.IteratorSize(::Type{<:WithTrailingZerosIterable}) = Base.SizeInfinite()


struct WithTrailingZerosVector <: AbstractVector{Int}
    data::Vector{Int}
end
Base.size(x::WithTrailingZerosVector) = (typemax(Int),)
Base.getindex(x::WithTrailingZerosVector, i::Int) = i <= length(x.data) ? x.data[i] : 0
function Base.show(io::IO, ::MIME"text/plain", x::WithTrailingZerosVector)
    println(io, "infinite-element WithTrailingZerosVector:")
    for i in 1:5
        x[i] >= 0 && print(io, ' ')
        println(io, x[i])
    end
    println(" ⋮")
end


function Base.collect(x::WithTrailingZerosIterable)
    data = Int[]
    while x !== nothing
        push!(data, x.head)
        x = x.tail
    end
    WithTrailingZerosVector(data)
end

function Base.sort!(x::WithTrailingZerosVector)
    filter!(x -> x < 0, x.data)
    sort!(x.data)
    return x
end


const X = WithTrailingZerosIterable(1, WithTrailingZerosIterable(-2, WithTrailingZerosIterable(3, nothing)))
display(collect(X))
#=
infinite-element WithTrailingZerosVector:
 1
-2
 3
 0
 0
 ⋮
=#

display(sort!(Base.copymutable(X)))
#=
infinite-element WithTrailingZerosVector:
-2
 0
 0
 0
 0
 ⋮
=#
```

Nevertheless, we could put this in the same camp as `sort(::AbstractString)`: perhaps it might make sense in some way but for now just throw because it is rarely a good idea to sort an infinite iterable.

Reply:
Throwing on an infinite iterator seems like the sensible choice. I've made the change.